## This Jupyter notebook answers HW12 questions for PHY 981 Nuclear Structure.

Author: Jacob Davison\
Date:   04/06/2021

### HW12 question 4

Use NuShellX to obtain the wavefunctions for 19O in the sd model space with
the usdb Hamiltonian for J up to 13/2. Are the results for the partitions
and maximum J value consistent with the answer to HW11.8?

The output I get from `shell` for 19O is:

`number of proton  partitions =            1
number of neutron partitions =            9
2*maxjp, 2*maxjn =            0          11`

The max Jn is definitely correct (11/2). Not sure about the number of partitions; in my homework I got 9 total neutron partitions and I think this is correct.

### HW12 question 5

Use NuShellX to obtain the wavefunctions for 19F in the 0d5/2 model space
and the usdb Hamiltonian for J up to 13/2. Are the J scheme dimensions
consistent with the answer to HW11.4? (This information is printed on the
screen and is in the \*.dim files.)

The output I get from `shell` for 19F is:

`\
approximate time to converge 10 states with 4-core (2.6 GHz)
 2J       J-dim         M-dim     t(min)   t(hour)
 13           1             1       0.20      0.00
 11           1             2       0.20      0.00
  9           2             4       0.20      0.00
  7           2             6       0.20      0.00
  5           3             9       0.20      0.00
  3           2            11       0.20      0.00
  1           1            12       0.20      0.00`
  
I create a `d5.mod` file with core O16 as the core in the 0d5/2 shell. Then, I run the `shell` with this model space on F19. The J-scheme dimensions are consistent with HW11.4, which include the isobaric analogue states of T=3/2.

### HW12 question 6

Modify the \*ans file from the previous problem to calculate the T values.
Are the (J; T) values in the \*.lpt file consistent with the answer to HW11.4?
Compare the calculated energy levels to experiment.

From `shell` for 19F adding the isospin contribution and reading the values in f19.lpt:

`\
  N    NJ   Ex(MeV)  pJ    T_z     lowest Ex     name      E(MeV)
    1    1    0.000   1/2  1/2    0.000     bb1301.lpe     -23.8610
    2    1    0.077   5/2  1/2    0.077     bb1305.lpe     -23.7837
    3    1    1.770   3/2  1/2    1.770     bb1303.lpe     -22.0906
    4    1    2.599   9/2  1/2    2.599     bb1309.lpe     -21.2624
    5    1    4.604  13/2  1/2    4.604     bb130d.lpe     -19.2572
    6    1    4.868   7/2  1/2    4.868     bb1307.lpe     -18.9934
    7    2    4.969   5/2  1/2              bb1305.lpe     -18.8916
    8    2    6.118   7/2  1/2              bb1307.lpe     -17.7435
    9    2    6.212   1/2  1/2              bb1301.lpe     -17.6485
   10    3    6.302   7/2  1/2              bb1307.lpe     -17.5590
   11    3    6.553   5/2  1/2              bb1305.lpe     -17.3084
   12    1    6.592  11/2  1/2    6.592     bb130b.lpe     -17.2692
`

I'm not sure how to read the number of allowed states from this, and in fact there are 128 rows in the output file which represent possible excited states (I assume). However, the rows that represent the lowest excited states are consistent with the possible (J,T) values in HW11.4.

I compute the T values here from the full `sd` model space. I can't get the (J,T) values to print using my model space.

### HW12 question 7

The Hamiltonian for a system of two basis states is: $H_{11} = -10$, $H_{12} =
-2$ and $H_{22} = -2$. Find the second-order perturbation correction to the
wavefunction and energy for state 1.

Second-order pertubative correction comes from the Taylor expansion of $\delta$ in Eq 31.4, written explicitly in Eq. 31.6.

Expand Eq. 31.4 in $H_{12} = 0$:

$$\delta = \frac{1}{2}\left[\sqrt{(H_{22} - H_{11})^2+4H_{H_{12}^2}}-(H_{22} - H_{11})\right] \approx \frac{1}{2}\left(H_{22} - H_{11} - H_{22} + H_{11}\right) + \frac{H_{12}^2}{(H_{22} - H_{11})}$$

Plug $\delta$ into $E_1$:

$$E_1 = H_{11} - \delta \approx H_{11} - \frac{H_{12}^2}{H_{22} - H_{11}} = -10 - \frac{4}{-2-(-10)} = -10.5$$

### HW12 question 8

In a different directory, obtain the (J; T) wavefunctions for 19F in the full
sd model space wth the usdb Hamiltonian. Compare the calculated energy
levels to experiment. How does the excitation energy of the lowest T = 3/2
state with experiment?

From `shell` running the full `sd` model on 19F:

`\
N    NJ   Ex(MeV)  pJ    T       lowest Ex     name      E(MeV)
1    1    0.000   1/2  1/2    0.000     bb1301.lpe     -23.8610
2    1    0.077   5/2  1/2    0.077     bb1305.lpe     -23.7837
3    1    1.770   3/2  1/2    1.770     bb1303.lpe     -22.0906
4    1    2.599   9/2  1/2    2.599     bb1309.lpe     -21.2624
5    1    4.604  13/2  1/2    4.604     bb130d.lpe     -19.2572
6    1    4.868   7/2  1/2    4.868     bb1307.lpe     -18.9934
7    2    4.969   5/2  1/2              bb1305.lpe     -18.8916
8    2    6.118   7/2  1/2              bb1307.lpe     -17.7435
9    2    6.212   1/2  1/2              bb1301.lpe     -17.6485
10    3    6.302   7/2  1/2              bb1307.lpe     -17.5590
11    3    6.553   5/2  1/2              bb1305.lpe     -17.3084
12    1    6.592  11/2  1/2    6.592     bb130b.lpe     -17.2692
13    2    6.819   3/2  1/2              bb1303.lpe     -17.0418
14    2    7.038   9/2  1/2              bb1309.lpe     -16.8225
15    3    7.080   3/2  1/2              bb1303.lpe     -16.7812
16    4    7.551   5/2  1/2              bb1305.lpe     -16.3105
17    5    7.905   5/2  3/2              bb1305.lpe     -15.9558`

Lowest excitation energies from shell model in 13/2 multiplet

|(J,T)| Shell model Ex (MeV) | NNDC Ex (MeV)|
|-----|----------------------|--------------|
|(5/2,1/2)|0.077|0.197|
|(3/2,1/2)|1.770|1.554|
|(9/2,1/2)|2.599|2.780|
|(13/2,1/2)|4.604|4.648|
|(7/2,1/2)|4.868|4.378|
|(11/2,1/2)|6.592|6.5|

The lowest T=3/2 according to shell model code is 7.905 MeV. Same state from experiment is 7.504 MeV according to NNDC.

### HW12 question 9

A simple model for the "pairing" part of the Hamiltonian is that all components of the wavefunction are connected by the same attractive interaction.
For example for three basis states, the two-body interaction part of the
matrix would look like

a a a\
a a a\
a a a

where a is the many-body interaction strength. Using a = -1 MeV, what
are the eigenvalues and wavefunctions for this matrix.

In [8]:
import numpy as np

H = np.array([[-1,-1,-1],
              [-1,-1,-1],
              [-1,-1,-1]])

w,v = np.linalg.eigh(H)
print(w)
print(v)

[-3.00000000e+00 -9.06674729e-18  4.53155957e-16]
[[-0.57735027  0.11435482  0.80844891]
 [-0.57735027  0.64295988 -0.50325864]
 [-0.57735027 -0.75731471 -0.30519027]]


The wavefunctions (eigenstates) expanded in basis states $\Phi_i,\quad i=1,2,3$ are:

$$
\begin{split}\Psi_1 &= -0.577(\Phi_1 + \Phi_2 + \Phi_3)\\
\Psi_2 &= 0.114\Phi_1 + 0.643\Phi_2 - 0.757\Phi_3\\
\Psi_3 &= 0.808\Phi_1 - 0.503\Phi_2 - 0.305\Phi_3
\end{split}
$$

The energies are:

$$E_1,E_2,E_3 = -3, -9.066\text{E-18}, 4.532\text{E-16}$$

Not sure if we can treat $E_2,E_3$ as zero, because of the distinct sign.

### HW12 question 10

For this last problem, find the ground state energy and wavefunction using
the Lanczos method. Start with the random vector (1; 2; 3)/$\sqrt{14}$

In [31]:
import numpy as np

H = np.array([[-1,-1,-1],
              [-1,-1,-1],
              [-1,-1,-1]])

# FIRST LANCZOS STEP
v_0 = np.array([1,2,3])/np.sqrt(14)
w_0 = H.dot(v_0)
a_0 = v_0.dot(w_0)
w_0 = w_0 - a_0*v_0
b_i = np.linalg.norm(w_0)
v_i = w_0/b_i

v_prev = v_0

# DO LANCZOS FOR N_ITERS
N_iters = 10
count = 1
while count < N_iters:

    w_i = H.dot(v_i)
    a_i = v_i.dot(w_i)
    b_i = v_prev.dot(w_i)
    
    v_prev = v_i
    
    w_i = w_i - a_i*v_i - b_i*v_prev
    v_i = w_i/np.linalg.norm(w_i)
    
    count += 1
    
    print('iter = {: 0d}\nE_1 = {: .8e}\nPsi_1 = ({: .8e},{: .8e},{: .8e})\n'.format(count,a_i, *v_i))

iter =  2
E_1 = -4.28571429e-01
Psi_1 = ( 8.06195636e-01, 5.32267823e-01, 2.58340010e-01)

iter =  3
E_1 = -2.54978132e+00
Psi_1 = (-2.56489420e-01,-5.31794047e-01,-8.07098674e-01)

iter =  4
E_1 = -2.54524418e+00
Psi_1 = ( 5.77124494e-01, 5.77350240e-01, 5.77575985e-01)

iter =  5
E_1 = -2.99999969e+00
Psi_1 = (-5.77369607e-01,-5.77350269e-01,-5.77330930e-01)

iter =  6
E_1 = -3.00000000e+00
Psi_1 = ( 5.77350269e-01, 5.77350269e-01, 5.77350269e-01)

iter =  7
E_1 = -3.00000000e+00
Psi_1 = (-5.77350269e-01,-5.77350269e-01,-5.77350269e-01)

iter =  8
E_1 = -3.00000000e+00
Psi_1 = ( 5.77350269e-01, 5.77350269e-01, 5.77350269e-01)

iter =  9
E_1 = -3.00000000e+00
Psi_1 = (-5.77350269e-01,-5.77350269e-01,-5.77350269e-01)

iter =  10
E_1 = -3.00000000e+00
Psi_1 = ( 5.77350269e-01, 5.77350269e-01, 5.77350269e-01)



Converges to the same results as question 9. After convergence, the sign of $\Psi_1$ flips back and forth because the orthogonal vector as already converged; that is, the next $v_{i+1}$ reduces to $v_{i+1} = v_i - 1*v_i - 1*v_{i-1} = -v_{i+1}$. However, the negative of a eigenvector of $H$ is still an eigenvector, so it doesn't matter.